sycl/doc/extensions/EnqueueBarrier/enqueue_barrier.asciidoc

```diff
@@ -0,0 +1,315 @@
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Contact
+Please open an issue in the https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/[extensions repository]
+
+== Dependencies
+
+This extension is written against the SYCL 1.2.1 specification, Revision v1.2.1-6.
+
+== Overview
+
+SYCL 1.2.1 defines a graph-based task execution model, based on kernels or explicit memory operations submitted to out-of-order queues. Dependencies between these kernels are represented by
+accessors that form data dependence edges in the execution graph. The USM extension <<usmlink,[1]>> doesn't have accessors, so instead solves
+this by defining `handler::depends_on` methods to specify event-based control dependencies between command groups.
+
+There are situations where defining dependencies based on events is more explicit than desired or required by an application. For instance, the user may know that a given task depends on all previously submitted tasks. Instead of explicitly adding all the required depends_on calls, the user could express this intent via a single call, making the program more concise and explicit.
+
+To simplify the interface, this extension adds an enqueued barrier that provides synchronization on SYCL command
+queues, with the following properties:
+
+1. Enqueued barriers are non-blocking from the host perspective.  The barrier is enqueued, so operates as part of the execution graph asynchronously from host program execution
+2. Command groups submitted to the same queue after the barrier is enqueued are not scheduled for execution until the barrier wait conditions are satisfied
+
+This proposal adds two new members to the `handler` class, and
+two new members to the `queue` class:
+
+[cols="70,70"]
+[grid="rows"]
+[options="header"]
+|========================================
+|*handler::barrier*|*queue::submit_barrier*
+|`void barrier()` | `event submit_barrier()`
+|`void barrier( const vector_class<event> &waitList )` | `event submit_barrier( const vector_class<event> &waitList )`
+|========================================
+
+The first variant of the barrier takes no parameters, and waits for all previously submitted commands to the queue to enter the `info::event_command_status::complete` state before any command later submitted to the same queue is allowed to execute. A second variant of the barrier accepts a list of events, with the behavior that no commands submitted to the same queue after barrier submission may execute until all events in the `waitList` have entered the `info::event_command_status::complete` state.  Both variants are non-blocking from the host program perspective, in that they do not wait for the barrier conditions to have been met before returning.
```

Comment:
What happens if the `submit_barrier` version that takes a `vector_class<event>` receives an empty list? does it act like the `submit_barrier` with no parameters, waiting for all submitted command groups to complete, or there is no wait at all? In that case, the event returned is immediately completed?

Reply:
This is described below in 4.8.X table:
  If `waitList` is empty, then the barrier has no effect.
Yes, the event returned is immediately completed.

Reply:
I think the behavior is fully defined on this.  If you disagree or want the material duplicated in this overview section, please comment.